keeper/msg_server_vaults_test.go

```diff
@@ -1890,3 +1890,143 @@ func TestFlexibleVaultRewardsEarlyExit(t *testing.T) {
 		},
 	}, rewards)
 }
+
+func TestVaultsProgramEnd(t *testing.T) {
+	account := mocks.AccountKeeper{
+		Accounts: make(map[string]sdk.AccountI),
+	}
+	bank := mocks.BankKeeper{
+		Balances: make(map[string]sdk.Coins),
+	}
+	k, _, ctx := mocks.DollarKeeperWithKeepers(t, bank, account)
+	bank.Restriction = k.SendRestrictionFn
+	k.SetBankKeeper(bank)
+
+	vaultsServer := keeper.NewVaultsMsgServer(k)
+	vaultsQueryServer := keeper.NewVaultsQueryServer(k)
+	bob, alice := utils.TestAccount(), utils.TestAccount()
+
+	// ARRANGE: Set the default index to 1.01 .
+	ctx = ctx.WithHeaderInfo(header.Info{Time: time.Date(2020, 1, 0, 0, 0, 0, 0, time.UTC)})
+	_ = k.UpdateIndex(ctx, 1.01e12)
+
+	// ARRANGE: Bob mints 1000 USDN.
+	_ = k.Mint(ctx, bob.Bytes, math.NewInt(1000*ONE), nil)
+
+	// ACT: Bob deposits 1000 USDN into the Staked Vault.
+	_, err := vaultsServer.Lock(ctx, &vaults.MsgLock{
+		Signer: bob.Address,
+		Vault:  vaults.VaultType(vaultsv1.VaultType_STAKED),
+		Amount: math.NewInt(1000 * ONE),
+	})
+	assert.NoError(t, err)
+
+	// ARRANGE: Increase the index from 1.0 to 1.1 (~10%).
+	ctx = ctx.WithHeaderInfo(header.Info{Time: time.Date(2020, 1, 1, 0, 0, 0, 0, time.UTC)})
+	_ = k.UpdateIndex(ctx, 1.1e12)
+
+	// ARRANGE: Bob mints 1000 USDN.
+	_ = k.Mint(ctx, bob.Bytes, math.NewInt(1000*ONE), nil)
+
+	// ACT: Bob deposits 1000 USDN into the Flexible Vault.
+	_, err = vaultsServer.Lock(ctx, &vaults.MsgLock{
+		Signer: bob.Address,
+		Vault:  vaults.VaultType(vaultsv1.VaultType_FLEXIBLE),
+		Amount: math.NewInt(1000 * ONE),
+	})
+	assert.NoError(t, err)
+
+	// ARRANGE: Increase the index from 1.1 to 1.21 (~10%).
+	_ = k.UpdateIndex(ctx, 1.21e12)
+	ctx = ctx.WithHeaderInfo(header.Info{Time: time.Date(2020, 1, 2, 0, 0, 0, 0, time.UTC)})
+
+	// ARRANGE: Alice mints 9000 USDN.
+	_ = k.Mint(ctx, alice.Bytes, math.NewInt(9000*ONE), nil)
+
+	// ACT: Alice deposits 9000 USDN into the Flexible Vault.
+	_, err = vaultsServer.Lock(ctx, &vaults.MsgLock{
+		Signer: alice.Address,
+		Vault:  vaults.VaultType(vaultsv1.VaultType_FLEXIBLE),
+		Amount: math.NewInt(9000 * ONE),
+	})
+	assert.NoError(t, err)
+
+	// ARRANGE: Increase the index from 1.21 to 1.33 (~10%).
+	_ = k.UpdateIndex(ctx, 1.33e12)
+	ctx = ctx.WithHeaderInfo(header.Info{Time: time.Date(2020, 1, 3, 0, 0, 0, 0, time.UTC)})
+
+	// ACT: Bob withdraws 1000 USDN from the Flexible Vault.
+	_, err = vaultsServer.Unlock(ctx, &vaults.MsgUnlock{
+		Signer: bob.Address,
+		Vault:  vaults.VaultType(vaultsv1.VaultType_FLEXIBLE),
+		Amount: math.NewInt(1000 * ONE),
+	})
+	assert.NoError(t, err)
+
+	// ARRANGE: Vaults Program Ends!
+	ctx = ctx.WithHeaderInfo(header.Info{Time: time.Date(2031, 1, 3, 0, 0, 0, 0, time.UTC)})
```

Comment:
nit: I think the tests would be more readable with some meaningful variable names instead of the random `time.Date(2031, 1, 3, ...)` and it would be less error-prone if the magic number in the mock keeper was made a constant `ExampleVaultsProgramEnd`


```go
	vaultsProgramEnd := time.Unix(mocks.ExampleVaultsProgramEnd, 0)
	afterVaultsProgramEnd := vaultsProgramEnd.Add(1)
```